builder/anka/step_clone_vm.go

```diff
@@ -0,0 +1,309 @@
+package anka
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/hashicorp/packer-plugin-sdk/multistep"
+	"github.com/hashicorp/packer-plugin-sdk/packer"
+	"github.com/veertuinc/packer-builder-veertu-anka/client"
+	"github.com/veertuinc/packer-builder-veertu-anka/common"
+	"github.com/veertuinc/packer-builder-veertu-anka/util"
+)
+
+// StepCloneVM will be used to run the clone step for any 'vm-clone' builder types
+type StepCloneVM struct {
+	client client.Client
+	vmName string
+}
+
+// Run clones a vm from a source vm either from an anka registry or locally
+func (s *StepCloneVM) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+	config := state.Get("config").(*Config)
+	ui := state.Get("ui").(packer.Ui)
+	ankaUtil := state.Get("util").(util.Util)
+	onError := func(err error) multistep.StepAction {
+		return ankaUtil.StepError(ui, state, err)
+	}
+	sourceVMTag := "latest"
+	doPull := config.AlwaysFetch
+
+	if config.SourceVMTag != "" {
+		sourceVMTag = config.SourceVMTag
+	}
+
+	s.client = state.Get("client").(client.Client)
+	s.vmName = config.VMName
+
+	if s.vmName == "" {
+		s.vmName = fmt.Sprintf("%s-%s", config.SourceVMName, ankaUtil.RandSeq(10))
+	}
+
+	state.Put("vm_name", s.vmName)
+
+	if config.PackerForce {
+		exists, err := s.client.Exists(s.vmName)
+		if err != nil {
+			return onError(err)
+		}
+		if exists {
+			ui.Say(fmt.Sprintf("Deleting existing virtual machine %s", s.vmName))
+
+			err = s.client.Delete(client.DeleteParams{VMName: s.vmName})
+			if err != nil {
+				return onError(err)
+			}
+		}
+	}
+
+	if !config.AlwaysFetch {
+		log.Printf("Searching for %s locally...", config.SourceVMName)
+
+		sourceExists, err := s.client.Exists(config.SourceVMName)
+		if err != nil {
+			return onError(err)
+		}
+		if !sourceExists {
+			log.Printf("Could not find %s locally, looking in anka registry...", config.SourceVMName)
+
+			doPull = true
+		}
+	}
+
+	if doPull {
+		ui.Say(fmt.Sprintf("Pulling source VM %s with tag %s from Anka Registry", config.SourceVMName, sourceVMTag))
+
+		registryParams := client.RegistryParams{
+			RegistryName: config.RegistryName,
+			RegistryURL:  config.RegistryURL,
+			NodeCertPath: config.NodeCertPath,
+			NodeKeyPath:  config.NodeKeyPath,
+			CaRootPath:   config.CaRootPath,
+			IsInsecure:   config.IsInsecure,
+		}
+
+		registryPullParams := client.RegistryPullParams{
+			VMID:   config.SourceVMName,
+			Tag:    sourceVMTag,
+			Local:  false,
```

Comment:
Nice! We can add a feature to allow local pulling only (avoiding a connection to the registry)